zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledStorageComponent.java

```diff
@@ -0,0 +1,179 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zipkin2.server.internal.throttle;
+
+import com.netflix.concurrency.limits.Limit;
+import com.netflix.concurrency.limits.limit.Gradient2Limit;
+import com.netflix.concurrency.limits.limiter.AbstractLimiter;
+import io.micrometer.core.instrument.MeterRegistry;
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import zipkin2.storage.SpanConsumer;
+import zipkin2.storage.SpanStore;
+import zipkin2.storage.StorageComponent;
+
+/**
+ * Delegating implementation that limits requests to the {@link #spanConsumer()} of another
+ * {@link StorageComponent}.  The theory here is that this class can be used to:
+ * <ul>
+ * <li>Prevent spamming the storage engine with excessive, spike requests when they come in; thus preserving it's life.</li>
+ * <li>Optionally act as a buffer so that a fixed number requests can be queued for execution when the throttle allows
+ *     for it.  This optional queue must be bounded in order to avoid running out of memory from infinitely queueing.</li>
+ * </ul>
+ *
+ * @see ThrottledSpanConsumer
+ */
+public final class ThrottledStorageComponent extends StorageComponent {
+  final StorageComponent delegate;
+  final AbstractLimiter<Void> limiter;
+  final ThreadPoolExecutor executor;
+
+  public ThrottledStorageComponent(StorageComponent delegate,
+                                   MeterRegistry registry,
+                                   int minConcurrency,
+                                   int maxConcurrency,
+                                   int maxQueueSize) {
+    this.delegate = Objects.requireNonNull(delegate);
+
+    Limit limit = Gradient2Limit.newBuilder()
+        .minLimit(minConcurrency)
+        .initialLimit(minConcurrency) // Limiter will trend towards min until otherwise necessary so may as well start there
+        .maxConcurrency(maxConcurrency)
+        .queueSize(0)
+        .build();
+    this.limiter = new Builder().limit(limit).build();
+
+    this.executor = new ThreadPoolExecutor(limit.getLimit(),
+                                           limit.getLimit(),
+                                           0,
+                                           TimeUnit.DAYS,
+                                           createQueue(maxQueueSize),
+                                           new ThottledThreadFactory(),
+                                           new ThreadPoolExecutor.AbortPolicy());
+    limit.notifyOnChange(new ThreadPoolExecutorResizer(executor));
+
+    if (registry != null) {
```

Comment:
do you expect this to be null?

Reply:
I don't, no.  But I don't know how the Spring injection stuff works and if the MeterRegistry can be "disabled" and therefore cause an NPE here.  If that's not possible I can not be defensive here but I'll have to lean on your judgement for that.

Reply:
Usually when something is disabled.. if it is also injected a bunch of places, that tool makes a no-op variant so that call sites don't need to inject anything. As this is our server, I don't anticipate us disabiling meters, and if we did and that would produce null, it would also break our other code.

One reason we don't support custom servers is to allow simpler decisions on topics like this.